src/Knp/Snappy/Chrome/AbstractGenerator.php

```diff
@@ -0,0 +1,160 @@
+     * @param array  $options    Set of options specific to this generation
+     *
+     * @throws \InvalidArgumentException When an invalid option is used
+     * @throws \RuntimeException         When backend fails to generate the output file
+     */
+    abstract protected function doGenerate(string $inputUri, string $outputFile, array $options);
+
+    abstract protected function getDefaultExtension(): string;
+
+    /**
+     * @return array
+     */
+    protected function getOptions(): array
+    {
+        return $this->options;
+    }
+
+    /**
+     * @return Backend
+     */
+    protected function getBackend(): Backend
+    {
+        return $this->backend;
+    }
+
+    /**
+     * @param Filesystem $filesystem
+     */
+    public function setFilesystem(Filesystem $filesystem)
+    {
+        $this->filesystem = $filesystem;
+    }
+
+    /**
+     * @param string $name
+     */
+    public function enableOption(string $name)
+    {
+        $this->options[$name] = true;
+    }
+
+    /**
+     * @param string $name
+     */
+    public function removeOption(string $name)
+    {
+        unset($this->options[$name]);
+    }
+
+    /**
+     * Set the default value for a specific option.
+     *
+     * @param string $name    Option name
+     * @param mixed  $default Default value
+     */
+    public function setOption(string $name, $default)
+    {
+        $this->options[$name] = $default;
+    }
+
+    /**
+     * Set default values for some options.
+     *
+     * @param array $options
+     */
+    public function setOptions(array $options)
+    {
+        $this->options = array_merge($this->options, $options);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function generate($input, string $output, array $options = [], bool $overwrite = false)
+    {
+        $this->filesystem->prepareOutput($output, $overwrite);
+
+        $this->doGenerate(sprintf('file://%s', $input), $output, $options);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function generateFromHtml($html, string $output, array $options = [], bool $overwrite = false)
+    {
+        $this->filesystem->prepareOutput($output, $overwrite);
+
+        $this->doGenerate(sprintf('data:text/html,%s', rawurlencode($html)), $output, $options);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getOutput($input, array $options = [])
+    {
+        $temporaryFile = $this->filesystem->createTemporaryFile(null, $this->getDefaultExtension());
+
+        $this->doGenerate(sprintf('file://%s', $input), $temporaryFile, $options);
+
+        return $this->filesystem->getFileContents($temporaryFile);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getOutputFromHtml($html, array $options = [])
+    {
+        $temporaryFile = $this->filesystem->createTemporaryFile(null, $this->getDefaultExtension());
+
+        $this->doGenerate(sprintf('data:text/html,%s', rawurlencode($html)), $temporaryFile, $options);
```

Comment:
Is there an advantage in doing this over a temp file? Performance doesn't seem to have a big impact here. Or would it be faster in slow-disk systems? It does have some limitations like 2MB limit (which is large for HTML, but still).

Also, rawurlencode doesn't seem to be required?